src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+{ idx_fun := λ i, (l.idx_fun i).map f,
+  proper  := ⟨l.proper.some, by rw [l.proper.some_spec, option.map_none'] ⟩ }
+
+/-- A point in `n → m` and a line in `n' → m` determine a line in `n ⊕ n' → m`. -/
+def vertical {m n n'} (v : n → m) (l : line m n') : line m (n ⊕ n') :=
+{ idx_fun := sum.elim (some ∘ v) l.idx_fun,
+  proper  := ⟨sum.inr l.proper.some, l.proper.some_spec⟩ }
+
+/-- A line in `n → m` and a point in `n' → m` determine a line in `n ⊕ n' → m`. -/
+def horizontal {m n n'} (l : line m n) (v : n' → m) : line m (n ⊕ n') :=
+{ idx_fun := sum.elim l.idx_fun (some ∘ v),
+  proper  := ⟨sum.inl l.proper.some, l.proper.some_spec⟩ }
+
+/-- One line in `n → m` and one in `n' → m` together determine a line in `n ⊕ n' → m`. -/
+def prod {m n n'} (l : line m n) (l' : line m n') : line m (n ⊕ n') :=
+{ idx_fun := sum.elim l.idx_fun l'.idx_fun,
+  proper  := ⟨sum.inl l.proper.some, l.proper.some_spec⟩ }
+
+lemma apply {m n} (l : line m n) (x : m) : l x = λ i, (l.idx_fun i).get_or_else x := rfl
+
+@[simp] lemma apply_none {m n} (l : line m n) (x : m) (i : n)
+  (h : l.idx_fun i = none) : l x i = x :=
+by simp only [option.get_or_else_none, h, l.apply]
+
+@[simp] lemma apply_of_ne_none {m n} (l : line m n) (x : m) (i : n)
+  (h : l.idx_fun i ≠ none) : some (l x i) = l.idx_fun i :=
+by rw [l.apply, option.get_or_else_of_ne_none h]
+
+@[simp] lemma map_apply {m m' n} (f : m → m') (l : line m n) (x : m) :
+  l.map f (f x) = f ∘ l x :=
+by simp only [line.apply, line.map, option.get_or_else_map]
+
+@[simp] lemma vertical_apply {m n n'} (v : n → m) (l : line m n') (x : m) :
+  l.vertical v x = sum.elim v (l x) :=
+by { funext i, cases i; refl }
+
+@[simp] lemma horizontal_apply {m n n'} (l : line m n) (v : n' → m) (x : m) :
+  l.horizontal v x = sum.elim (l x) v :=
+by { funext i, cases i; refl }
+
+@[simp] lemma prod_apply {m n n'} (l : line m n) (l' : line m n') (x : m) :
+  l.prod l' x = sum.elim (l x) (l' x) :=
+by { funext i, cases i; refl }
+
+/-- The Hales-Jewett theorem: for any finite types `m` and `k` with `m` nonempty, there exists
+another finite type `n` such that whenever the hypercube `n → m` is `k`-colored, there is
+a monochromatic combinatorial line. -/
+theorem exists_mono_in_high_dimension : ∀ m [fintype m] [nonempty m] k [fintype k],
+  ∃ n [fintype n], ∀ C : (n → m) → k, ∃ l : line m n, l.is_mono C :=
+-- The proof proceeds by induction on `m`.
+fintype_induction
+-- We have to show that the theorem is invariant under `m ≃ m'` for the induction to work.
+(λ m m' e, forall_imp $ λ k, forall_imp $ λ _, Exists.imp $ λ n, Exists.imp $ λ _ h C,
+  let ⟨l, c, lc⟩ := h (λ v, C (e ∘ v)) in
+  ⟨l.map e, c, e.forall_congr_left.mp $ λ x, by rw [←lc x, line.map_apply]⟩)
+-- The base case `m = unit` is easy since any line is monochromatic.
+(λ k _, ⟨unit, infer_instance, λ C, ⟨diagonal _ _, C (λ _, unit.star), λ ⟨⟩, rfl⟩⟩)
+begin -- Now we have to show that the theorem holds for `unit m` if it holds for `m`.
```

Comment:
```suggestion
begin -- Now we have to show that the theorem holds for `option m` if it holds for `m`.
```

Reply:
is what I think you meant?

Reply:
Yep, thanks! I get these mixed up because of `option x = m + unit`...